pkg/controllers/nodeclass/instanceprofile.go

```diff
@@ -29,30 +31,84 @@ import (
 type InstanceProfile struct {
 	instanceProfileProvider instanceprofile.Provider
 	region                  string
+	recreationCache         *cache.Cache
 }
 
-func NewInstanceProfileReconciler(instanceProfileProvider instanceprofile.Provider, region string) *InstanceProfile {
+func NewInstanceProfileReconciler(instanceProfileProvider instanceprofile.Provider, region string, cache *cache.Cache) *InstanceProfile {
 	return &InstanceProfile{
 		instanceProfileProvider: instanceProfileProvider,
 		region:                  region,
+		recreationCache:         cache,
 	}
 }
 
+func (ip *InstanceProfile) PreventRecreation(nodeClass *v1.EC2NodeClass) (string, bool) {
+	if profileName, found := ip.recreationCache.Get(fmt.Sprintf("%s/%s", nodeClass.Spec.Role, nodeClass.UID)); found {
+		return profileName.(string), true
+	}
+	return "", false
+}
+
 func (ip *InstanceProfile) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass) (reconcile.Result, error) {
 	if nodeClass.Spec.Role != "" {
-		profileName := nodeClass.InstanceProfileName(options.FromContext(ctx).ClusterName, ip.region)
-		if err := ip.instanceProfileProvider.Create(
-			ctx,
-			profileName,
-			nodeClass.InstanceProfileRole(),
-			nodeClass.InstanceProfileTags(options.FromContext(ctx).ClusterName, ip.region),
-		); err != nil {
-			return reconcile.Result{}, fmt.Errorf("creating instance profile, %w", err)
+		var currentRole string
+		var oldProfileName string
+
+		// Use a short-lived cache to prevent instance profile recreation for the same role in the same EC2NodeClass
+		// in case of a status patch error in the EC2NodeClass controller
+		if profileName, found := ip.PreventRecreation(nodeClass); found {
```

Comment:
nit: `ok` is the universal variable name for existence checks
```suggestion
		if profileName, ok := ip.PreventRecreation(nodeClass); ok {
```